File: app/src/main/java/com/mainstreetcode/teammate/adapters/viewholders/input/InputViewHolder.java
package com.mainstreetcode.teammate.adapters.viewholders.input;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.baseclasses.BaseViewHolder;
import com.mainstreetcode.teammate.fragments.headless.ImageWorkerFragment;
import com.mainstreetcode.teammate.model.Item;
import com.mainstreetcode.teammate.util.ErrorHandler;

import java.util.Objects;

import androidx.annotation.Nullable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

import static android.text.TextUtils.isEmpty;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static androidx.core.content.ContextCompat.getColor;
import static androidx.core.content.ContextCompat.getDrawable;
import static com.mainstreetcode.teammate.util.ViewHolderUtil.listenForLayout;
import static io.reactivex.android.schedulers.AndroidSchedulers.mainThread;

public class InputViewHolder<T extends ImageWorkerFragment.ImagePickerListener> extends BaseViewHolder<T>
        implements
        TextWatcher {

    private static final int HINT_ANIMATION_DURATION = 200;
    private static final float HINT_SHRINK_SCALE = 0.8F;
    private static final float HALF = 0.5F;

    private int lastLineCount = 1;

    protected final TextView hint;
    protected final EditText text;
    private final ImageButton button;

    @Nullable
    TextInputStyle textInputStyle;

    public InputViewHolder(View itemView) {
        super(itemView);
        hint = itemView.findViewById(R.id.hint);
        text = itemView.findViewById(R.id.input);
        button = itemView.findViewById(R.id.button);
        text.setOnFocusChangeListener((v, hasFocus) -> {
            tintHint(hasFocus);
            scaleHint(!hasFocus && isEmpty(text.getText()));
        });
    }

    @Override protected void clear() {
        button.setOnClickListener(null);
        text.setOnClickListener(null);
        text.removeTextChangedListener(this);

        if (textInputStyle != null) textInputStyle.setViewHolder(null);
        textInputStyle = null;

        super.clear();
    }

    @Override protected void onDetached() {
        button.setVisibility(GONE);
        super.onDetached();
    }

    protected float getHintLateralTranslation() {
        int width = hint.getWidth();
        return -((width - (HINT_SHRINK_SCALE * width)) * HALF);
    }

    protected float getHintLongitudinalTranslation() {
        return -((itemView.getHeight() - hint.getHeight()) * HALF);
    }

    public void bind(TextInputStyle inputStyle) {
        this.textInputStyle = inputStyle;
        inputStyle.setViewHolder(this);

        Item item = inputStyle.getItem();

        int newInputType = item.getInputType();
        int oldInputType = text.getInputType();

        boolean isEditable = inputStyle.isEditable();
        boolean isSelector = inputStyle.isSelector();

        boolean isEnabled = text.isEnabled();
        boolean isClickable = text.isClickable();
        boolean isFocusable = text.isFocusable();
        boolean isFocusableInTouchMode = text.isFocusableInTouchMode();

        CharSequence newValue = item.getValue().toString();
        CharSequence oldValue = text.getText().toString();

        CharSequence oldHint = hint.getText();
        CharSequence newHint = itemView.getContext().getString(item.getStringRes());

        if (isEnabled != isEditable) text.setEnabled(isEditable);
        if (isClickable != isSelector) text.setClickable(isSelector);
        if (isFocusable == isSelector) text.setFocusable(!isSelector);
        if (isFocusableInTouchMode == isSelector) text.setFocusableInTouchMode(!isSelector);

        if (!Objects.equals(oldHint, newHint)) hint.setText(newHint);
        if (!Objects.equals(oldValue, newValue)) text.setText(newValue);
        if (oldInputType != newInputType) text.setInputType(newInputType);

        text.setOnClickListener(inputStyle.textClickListener());
        button.setOnClickListener(inputStyle.buttonClickListener());

        text.removeTextChangedListener(this);
        if (!isSelector) text.addTextChangedListener(this);

        updateButton();
        checkForErrors();
        setTintAlpha(text.hasFocus());
        listenForLayout(hint, () -> scaleHint(isEmpty(text.getText())));
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {/* Nothing */}

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {/* Nothing */}

    @Override
    public void afterTextChanged(Editable editable) {
        int currentLineCount = text.getLineCount();
        if (lastLineCount != currentLineCount) listenForLayout(hint, () -> scaleHint(false));
        lastLineCount = currentLineCount;

        if (textInputStyle == null || textInputStyle.isSelector()) return;
        textInputStyle.getItem().setValue(editable.toString());
        checkForErrors();
    }

    void updateText(CharSequence text) {
        this.text.setText(text);
        checkForErrors();
        listenForLayout(hint, () -> scaleHint(isEmpty(text)));
    }

    private void checkForErrors() {
        if (textInputStyle == null) return;

        CharSequence errorMessage = textInputStyle.errorText();
        if (Objects.equals(errorMessage, text.getError())) return;

        if (isEmpty(errorMessage)) text.setError(null);
        else text.setError(errorMessage);
    }

    private void updateButton() {
        if (textInputStyle == null) return;

        int newIcon = textInputStyle.getIcon();

        int oldVisibility = button.getVisibility();
        int newVisibility = newIcon == 0 ? GONE : VISIBLE;

        if (oldVisibility != newVisibility) button.setVisibility(newVisibility);
        if (newIcon != 0)
            disposables.add(Single.fromCallable(() -> getDrawable(text.getContext(), newIcon))
                    .subscribeOn(Schedulers.io())
                    .observeOn(mainThread())
                    .subscribe(button::setImageDrawable, ErrorHandler.EMPTY));
    }

    private void scaleHint(boolean grow) {
        float scale = grow ? 1F : HINT_SHRINK_SCALE;
        float translationX = grow ? 0 : getHintLateralTranslation();
        float translationY = grow ? 0 : getHintLongitudinalTranslation();

        hint.animate()
                .scaleX(scale)
                .scaleY(scale)
                .translationX(translationX)
                .translationY(translationY)
                .setDuration(HINT_ANIMATION_DURATION)
                .start();
    }

    private void tintHint(boolean hasFocus) {
        int start = hint.getCurrentTextColor();
        int end = getColor(hint.getContext(), hasFocus ? R.color.colorAccent : R.color.dark_grey);

        ValueAnimator animator = ValueAnimator.ofObject(new ArgbEvaluator(), start, end);
        animator.setDuration(HINT_ANIMATION_DURATION);
        animator.addUpdateListener(animation -> hint.setTextColor((int) animation.getAnimatedValue()));
        animator.addListener(new AnimatorListenerAdapter() {
            @Override public void onAnimationEnd(Animator animation) { setTintAlpha(hasFocus); }
        });
        animator.start();
    }

    private void setTintAlpha(boolean hasFocus) {
        hint.setAlpha(textInputStyle != null && !textInputStyle.isEditable() && !hasFocus ? 0.38F : 1F);
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/adapters/viewholders/input/InputViewHolder.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.adapters.viewholders.input

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.EditText
import android.widget.ImageButton
import android.widget.TextView
import androidx.core.content.ContextCompat.getDrawable
import androidx.core.view.doOnLayout
import androidx.recyclerview.widget.RecyclerView
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.util.ErrorHandler
import com.mainstreetcode.teammate.util.resolveThemeColor
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

open class InputViewHolder(
        itemView: View
) : RecyclerView.ViewHolder(itemView), TextWatcher {

    private val disposables = CompositeDisposable()

    private var lastLineCount = 1

    protected val hint: TextView = itemView.findViewById(R.id.hint)
    private val textField: EditText = itemView.findViewById(R.id.input)
    private val button: ImageButton = itemView.findViewById(R.id.button)

    internal var textInputStyle: TextInputStyle? = null

    protected open val hintLateralTranslation: Float
        get() {
            val width = hint.width
            return -((width - HINT_SHRINK_SCALE * width) * HALF)
        }

    protected open val hintLongitudinalTranslation: Float
        get() = -((itemView.height - hint.height) * HALF)

    init {
        textField.setOnFocusChangeListener { _, hasFocus ->
            tintHint(hasFocus)
            scaleHint(!hasFocus && textField.text.isBlank())
        }
    }

    fun clear() {
        button.setOnClickListener(null)
        textField.setOnClickListener(null)
        textField.removeTextChangedListener(this)

        textInputStyle?.viewHolder = null
        textInputStyle = null

        disposables.clear()
    }

    fun onDetached() {
        button.visibility = GONE
    }

    open fun bind(inputStyle: TextInputStyle) {
        this.textInputStyle = inputStyle
        inputStyle.viewHolder = this

        val item = inputStyle.item

        val newInputType = item.inputType
        val oldInputType = textField.inputType

        val isEditable = inputStyle.isEditable
        val isSelector = inputStyle.isSelector

        val isEnabled = textField.isEnabled
        val isClickable = textField.isClickable
        val isFocusable = textField.isFocusable
        val isFocusableInTouchMode = textField.isFocusableInTouchMode

        val newValue = item.formattedValue.toString()
        val oldValue = textField.text.toString()

        val oldHint = hint.text
        val newHint = itemView.context.getString(item.stringRes)

        if (isEnabled != isEditable) textField.isEnabled = isEditable
        if (isClickable != isSelector) textField.isClickable = isSelector
        if (isFocusable == isSelector) textField.isFocusable = !isSelector
        if (isFocusableInTouchMode == isSelector) textField.isFocusableInTouchMode = !isSelector

        if (oldHint != newHint) hint.text = newHint
        if (oldValue != newValue) textField.setText(newValue)
        if (oldInputType != newInputType) textField.inputType = newInputType

        textField.setOnClickListener(inputStyle.textClickListener())
        button.setOnClickListener(inputStyle.buttonClickListener())

        textField.removeTextChangedListener(this)
        if (!isSelector) textField.addTextChangedListener(this)

        updateButton()
        checkForErrors()
        setTintAlpha(textField.hasFocus())
        hint.doOnLayout { scaleHint(textField.text.isBlank()) }
    }

    override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) = Unit /* Nothing */

    override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) = Unit /* Nothing */

    override fun afterTextChanged(editable: Editable) {
        val currentLineCount = textField.lineCount
        if (lastLineCount != currentLineCount) hint.doOnLayout { scaleHint(false) }
        lastLineCount = currentLineCount

        if (textInputStyle == null || textInputStyle!!.isSelector) return
        textInputStyle?.item?.rawValue = editable.toString()
        checkForErrors()
    }

    internal open fun updateText(text: CharSequence) {
        this.textField.setText(text)
        checkForErrors()
        hint.doOnLayout { scaleHint(text.isBlank()) }
    }

    private fun checkForErrors() {
        val errorMessage = textInputStyle?.errorText()
        if (errorMessage == textField.error) return

        textField.error = if (errorMessage.isNullOrBlank()) null else errorMessage
    }

    private fun updateButton() {
        if (textInputStyle == null) return

        val newIcon = textInputStyle!!.icon

        val oldVisibility = button.visibility
        val newVisibility = if (newIcon == 0) GONE else VISIBLE

        if (oldVisibility != newVisibility) button.visibility = newVisibility
        if (newIcon != 0)
            disposables.add(Single.fromCallable { getDrawable(textField.context, newIcon) }
                    .subscribeOn(Schedulers.io())
                    .observeOn(mainThread())
                    .subscribe(button::setImageDrawable, ErrorHandler.EMPTY::invoke))
    }

    private fun scaleHint(grow: Boolean) {
        val scale = if (grow) 1F else HINT_SHRINK_SCALE
        val translationX = if (grow) 0F else hintLateralTranslation
        val translationY = if (grow) 0F else hintLongitudinalTranslation

        hint.animate()
                .scaleX(scale)
                .scaleY(scale)
                .translationX(translationX)
                .translationY(translationY)
                .setDuration(HINT_ANIMATION_DURATION.toLong())
                .start()
    }

    private fun tintHint(hasFocus: Boolean) {
        val start = hint.currentTextColor
        val end = hint.context.resolveThemeColor(if (hasFocus) R.attr.colorSecondary else R.attr.main_text_color)

        ValueAnimator.ofObject(ArgbEvaluator(), start, end).apply {
            duration = HINT_ANIMATION_DURATION.toLong()
            addUpdateListener { animation -> hint.setTextColor(animation.animatedValue as Int) }
            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) = setTintAlpha(hasFocus)
            })
            start()
        }
    }

    private fun setTintAlpha(hasFocus: Boolean) {
        val notEditable = textInputStyle?.isEditable?.not() ?: false
        hint.alpha = if (notEditable && !hasFocus) 0.38f else 1f
    }

    companion object {

        private const val HINT_ANIMATION_DURATION = 200
        private const val HINT_SHRINK_SCALE = 0.8f
        private const val HALF = 0.5f
    }
}
